pkg/segment/aggregations/segaggs.go

```diff
@@ -196,6 +209,106 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
 	return nil
 }
 
+func performTail(nodeResult *structs.NodeResult, aggs *structs.QueryAggregators, tailExpr *structs.TailExpr, recs map[string]map[string]interface{}, recordIndexInFinal map[string]int, finishesSegment bool, numTotalSegments uint64, hasSort bool) error {
+
+	if nodeResult.Histogram != nil {
+		for _, aggResult := range nodeResult.Histogram {
+			diff := len(aggResult.Results) - int(aggs.OutputTransforms.TailRows)
+			if diff > 0 {
+				aggResult.Results = aggResult.Results[diff:]
+				aggs.OutputTransforms.TailRows = 0
+			} else {
+				aggs.OutputTransforms.TailRows -= uint64(len(aggResult.Results))
+			}
+			n := len(aggResult.Results)
+			for i := 0; i < n/2; i++ {
+				aggResult.Results[i], aggResult.Results[n-i-1] = aggResult.Results[n-i-1], aggResult.Results[i]
+			}
+			if aggs.OutputTransforms.TailRows == 0 {
+				break
+			}
+		}
+
+		return nil
+	}
+
+	if finishesSegment {
+		tailExpr.NumProcessedSegments++
+	}
+
+	if tailExpr.TailRecords == nil {
+		tailExpr.TailRecords = make(map[string]map[string]interface{}, 0)
+	}
+	if tailExpr.TailPQ == nil {
+		pq := make(utils.PriorityQueue, 0)
+		tailExpr.TailPQ = &pq
+		heap.Init(tailExpr.TailPQ)
+	}
+
+	if !hasSort {
+		for k, v := range recs {
+			timeVal, exists := v["timestamp"]
```

Comment:
nit: rename `k` and `v`. I'm not sure what `k` is, but it looks like `v` is a single record so maybe renaming it to `record` would make sense

Reply:
addressed in the latest commit. k refers to the string key use to map actual record in recs